sdk-actors/src/main/java/io/dapr/actors/client/DaprGrpcClient.java

```diff
@@ -5,59 +5,126 @@
+import reactor.core.publisher.MonoSink;
+import reactor.util.context.Context;
+
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
 
 /**
  * A DaprClient over GRPC for Actor.
  */
 class DaprGrpcClient implements DaprClient {
 
   /**
-   * The GRPC client to be used.
-   *
-   * @see DaprGrpc.DaprFutureStub
+   * The async gRPC stub.
    */
-  private DaprGrpc.DaprFutureStub client;
+  private DaprGrpc.DaprStub client;
 
   /**
    * Internal constructor.
    *
    * @param grpcClient Dapr's GRPC client.
    */
-  DaprGrpcClient(DaprGrpc.DaprFutureStub grpcClient) {
-    this.client = grpcClient;
+  DaprGrpcClient(DaprGrpc.DaprStub grpcClient) {
+    this.client = intercept(grpcClient);
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public Mono<byte[]> invoke(String actorType, String actorId, String methodName, byte[] jsonPayload) {
-    return Mono.fromCallable(DaprException.wrap(() -> {
-      DaprProtos.InvokeActorRequest req =
-          DaprProtos.InvokeActorRequest.newBuilder()
-              .setActorType(actorType)
-              .setActorId(actorId)
-              .setMethod(methodName)
-              .setData(jsonPayload == null ? ByteString.EMPTY : ByteString.copyFrom(jsonPayload))
-              .build();
-
-      return get(client.invokeActor(req));
-    })).map(r -> r.getData().toByteArray());
+    DaprProtos.InvokeActorRequest req =
+        DaprProtos.InvokeActorRequest.newBuilder()
+            .setActorType(actorType)
+            .setActorId(actorId)
+            .setMethod(methodName)
+            .setData(jsonPayload == null ? ByteString.EMPTY : ByteString.copyFrom(jsonPayload))
+            .build();
+    return Mono.subscriberContext().flatMap(
+        context -> this.<DaprProtos.InvokeActorResponse>createMono(
+            it -> intercept(context, client).invokeActor(req, it)
+        )
+    ).map(r -> r.getData().toByteArray());
+  }
+
+  /**
+   * Populates GRPC client with interceptors.
+   *
+   * @param client GRPC client for Dapr.
+   * @return Client after adding interceptors.
+   */
+  private static DaprGrpc.DaprStub intercept(DaprGrpc.DaprStub client) {
+    ClientInterceptor interceptor = new ClientInterceptor() {
+      @Override
+      public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+          MethodDescriptor<ReqT, RespT> methodDescriptor,
+          CallOptions callOptions,
+          Channel channel) {
+        ClientCall<ReqT, RespT> clientCall = channel.newCall(methodDescriptor, callOptions);
+        return new ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(clientCall) {
+          @Override
+          public void start(final Listener<RespT> responseListener, final Metadata metadata) {
+            String daprApiToken = Properties.API_TOKEN.get();
+            if (daprApiToken != null) {
+              metadata.put(Metadata.Key.of("dapr-api-token", Metadata.ASCII_STRING_MARSHALLER), daprApiToken);
```

Comment:
Is there a negative case if the token is empty? Or is this related to the feature that has to be enabled and if it's not we hit the null (expected) case.

Reply:
We hit null if not set. If it is set as empty, it is OK to pass it through.